controlplane/kubeadm/internal/control_plane.go

```diff
@@ -100,11 +100,12 @@ func (c *ControlPlane) MachinesNeedingUpgrade() FilterableMachineCollection {
 		return c.Machines.AnyFilter(
 			machinefilters.Not(machinefilters.MatchesConfigurationHash(c.SpecHash())),
 			machinefilters.OlderThan(c.KCP.Spec.UpgradeAfter),
-		)
+		).Filter(machinefilters.Not(machinefilters.HasDeletionTimestamp))
 	}
 
 	return c.Machines.Filter(
 		machinefilters.Not(machinefilters.MatchesConfigurationHash(c.SpecHash())),
+		machinefilters.Not(machinefilters.HasDeletionTimestamp),
```

Comment:
cc @benmoss @detiber

Reply:
This can be added in scaleUpControlPlane() similar to scaleDownControlPlane():
```
	// Wait for any delete in progress to complete before deleting another Machine
	if controlPlane.HasDeletingMachine() {
		return ctrl.Result{}, &capierrors.RequeueAfterError{RequeueAfter: deleteRequeueAfter}
	}
```
I prefer to do it in MachinesNeedingUpgrade(), because Machines that are about to get deleted should not be included in that list logically.